nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py

```diff
@@ -1156,91 +1216,127 @@ def dummy():
+            log_probs = self._one_step_forward(
+                enc_output,
+                enc_output_attn_mask,
+                prefixes,
+                prefixes != tokenizer.pad_id,
+                tensor_shape,
+                tokenizer.vocab_size,
+                device,
+                ignore_ids,
+            )
+
+            # get top candidates for each item in batch
+            scores_i, prefixes_i = torch.topk(log_probs[:, -1, :], beam_size, dim=-1)
+
+            # mask all finished hypotheses to exclude them from beam
+            pad_mask = pad_profile.repeat(1, beam_size)
+
+            # for all prefixes ending with <eos> or <pad> replace generated
+            # continuations with <pad>
+            prefixes_i = tokenizer.pad_id * pad_mask + prefixes_i * (1 - pad_mask)
+
+            # force all hypotheses but one generated from already finished
+            # hypotheses to have extremely low score, so they will not be
+            # considered during beam re-ranking
+            pad_mask[:, 1:] = pad_mask[:, 1:] * -10000.0
+            scores = scores + scores_i * (1 - pad_mask).to(scores.dtype)
+
+            # choose top-k hypotheses with length penalty applied
+            len_penalties = self.compute_len_penalty(prefixes_len, beam_alpha)
+            scores = scores / len_penalties
+            scores, indices_i = torch.topk(scores.view(-1, beam_size ** 2), beam_size, dim=1)
+            scores = scores.view(-1, 1) * len_penalties
+
+            # select prefixes which correspond to the chosen hypotheses
+            prefixes = prefixes.unsqueeze(1).repeat(1, beam_size, 1)
+            prefixes = torch.cat((prefixes, prefixes_i.unsqueeze(2)), dim=2)
+            prefixes = prefixes.view(batch_size, beam_size ** 2, -1)
+            p_len = prefixes.size(2)
+            prefixes_ids = indices_i.unsqueeze(2).repeat(1, 1, p_len)
+            prefixes = prefixes.gather(1, prefixes_ids).view(-1, p_len)
+
+            # select logits similar to prefixes
+            all_logits = all_logits.unsqueeze(1).repeat(1, beam_size, 1)
+            all_logits = torch.cat((all_logits, scores_i.unsqueeze(2)), dim=2)
+            all_logits = all_logits.view(batch_size, beam_size ** 2, -1)
+            all_logits = all_logits.gather(1, prefixes_ids).view(-1, p_len)
+
+            # update prefixes_len and pad_profile
+            not_eos_pad = prefixes.ne(tokenizer.eos_id) & prefixes.ne(tokenizer.pad_id)
+            prefixes_len = 1 + not_eos_pad.sum(dim=1, keepdim=True).to(scores.dtype)
+            pad_profile = (~not_eos_pad[:, -1:]).long()
+
+            # if all hypotheses end with <eos> or <pad>, interrupt search
+            if pad_profile.sum() == batch_size * beam_size:
+                break
+
+        # select best performing hypotheses in each element of the batch
+        len_penalties = self.compute_len_penalty(prefixes_len, 0)
+        scores = scores / len_penalties
+        best_tokens = (
```

Comment:
I think it would be useful to leave here an option for the user to provide all topK predictions. At least it is applicable in my case.  It can be achieved via this if statement

```
  if beam_size == 1 or keep_only_best_tokens:
      best_ids = torch.argmax(scores.view(-1, beam_size), dim=1, keepdim=True)
      scores = scores.view(batch_size, beam_size).gather(1, best_ids)
      best_tokens = best_ids.repeat(1, prefixes.size(1)).unsqueeze(1)
      tgt_text = prefixes.view(batch_size, beam_size, -1).gather(1, best_tokens).squeeze(1)
      logits = all_logits.view(batch_size, beam_size, -1).gather(1, best_tokens).squeeze(1)
  else:
      tgt_text = prefixes.view(batch_size, beam_size, -1)
      logits = all_logits.view(batch_size, beam_size, -1)
```